### PR TITLE
Replace platform detection macros with simple `final val` definitions…

### LINKED ISSUE #22: Consider using final vals for Platform.{isJvm,isJs}
Just browsing through the code, I noticed you use macros to implement `Platform.isJvm` and `Platform.isJs`. The macros are different on both platforms and return trivial constant expressions: [JVM version](https://github.com/InTheNow/catalysts/blob/master/platform/jvm/src/main/scala/catalysts/Platform_jvm.scala) and [JS version](https://github.com/InTheNow/catalysts/blob/master/platform/js/src/main/scala/catalysts/Platform_js.scala).

You can achieve precisely the same result without macros, with `final val`s without type ascription. JVM version:
```scala
object Platform {
  final val isJvm = true
  final val isJs = false
}
```
and you guess the JS version.

final vals without type ascription whose right-hand-side is a literal, or can be constant-folded away, are themselves constant-folded at use site by `typer`, so you get exactly the same tree after typer as if you use the macros.

### PR DESCRIPTION
**Pull request by ochrons**
… that the compiler can constant-fold at call site. Fixes #22.

### CODE REVIEW

File: tests/src/test/scala/catalysts/tests/PlatformTests.scala [hunk 1]
```diff
@@ -21,5 +21,14 @@ class PlatformTests extends TestSuite {
     assert(Platform.isJs != Platform.isJvm)
     assert(hackyIsJS != hackyIsJVM)
   }
-}
 
+  test("dead code elimination") {
+    // look at disassembled class files to verify this reduces to a `bipush 42` or `bipush 43`
```
Comment (InTheNow) [inline tests/src/test/scala/catalysts/tests/PlatformTests.scala]: nice!

File: tests/src/test/scala/catalysts/tests/PlatformTests.scala [hunk 2]
```diff
@@ -21,5 +21,14 @@ class PlatformTests extends TestSuite {
...
+    val x = if( Platform.isJvm ) {
```
Comment (InTheNow) [inline tests/src/test/scala/catalysts/tests/PlatformTests.scala]: perhaps just a one liner, no curlies?

**Comment by InTheNow** [comment]
I think that because the macros have been removed that macro-paradise can be [removed from the build](https://github.com/InTheNow/catalysts/blob/master/build.sbt#L107)

otherwise :+1:

**Comment by InTheNow** [comment]
LGTM :+1:

**Comment by InTheNow** [comment]
@non As this has is _quite_ a large change on you original idea, could you review as well, please?

**Comment by non** [comment]
Looks good to me. :+1:

In the long run if we decided we didn't want to rely on constant-folding + branch-elimination we could build different macros to do it explicitly. But for now this seems great.

**Comment by sjrd** [comment]
One way to assert for certain that branch-elimination is applied would be to have a test like this:
```scala
if (Platform.isJvm)
  Thread.sleep(1)
```
That test would fail to *link* on the JS platform if the branch is not eliminated.

**Comment by InTheNow** [comment]
@sjrd - Thanks for these insightful suggestions (although I'm pretty sure @ochrons would have already mused over this one :wink: ). I'll pop in this test in next PR, unless I'm beaten to it.

**Pull request merged**